plugins/usbpro/OpenDeckDevice.h

```diff
@@ -0,0 +1,131 @@
+#include "olad/TokenBucket.h"
+#include "olad/PluginAdaptor.h"
+#include "olad/Port.h"
+
+#include "plugins/usbpro/OpenDeckWidget.h"
+#include "plugins/usbpro/UsbSerialDevice.h"
+#include "plugins/usbpro/messages/UsbProConfigMessages.pb.h"
+
+namespace ola {
+namespace plugin {
+namespace usbpro {
+
+/*
+ * An OpenDeck device
+ */
+class OpenDeckDevice: public UsbSerialDevice {
+ public:
+  OpenDeckDevice(ola::PluginAdaptor *plugin_adaptor,
+                 ola::AbstractPlugin *owner,
+                 const std::string &name,
+                 OpenDeckWidget *widget,
+                 uint16_t esta_id,
+                 uint16_t device_id,
+                 uint32_t serial,
+                 uint16_t firmware_version,
+                 unsigned int fps_limit);
+
+  std::string DeviceId() const { return m_serial; }
+
+  void Configure(ola::rpc::RpcController *controller,
+                 const std::string &request,
+                 std::string *response,
+                 ConfigureCallback *done);
+
+ protected:
+  void PrePortStop();
+
+ private:
+  void UpdateParams(bool status, const usb_pro_parameters &params);
+
+  void HandleParametersRequest(ola::rpc::RpcController *controller,
+                               const ola::plugin::usbpro::Request *request,
+                               std::string *response,
+                               ConfigureCallback *done);
+
+  void HandleParametersResponse(ola::rpc::RpcController *controller,
+                                std::string *response,
+                                ConfigureCallback *done,
+                                bool status,
+                                const usb_pro_parameters &params);
+
+  void HandleSerialRequest(ola::rpc::RpcController *controller,
+                           const ola::plugin::usbpro::Request *request,
+                           std::string *response,
+                           ConfigureCallback *done);
+
+  OpenDeckWidget *m_opendeck_widget;
+  std::string m_serial;
+
+  bool m_got_parameters;
+  uint8_t m_break_time;
+  uint8_t m_mab_time;
+  uint8_t m_rate;
+};
+
+/*
+ * The output port
+ */
+class OpenDeckOutputPort: public BasicOutputPort {
+ public:
+  OpenDeckOutputPort(OpenDeckDevice *parent,
+                     OpenDeckWidget *widget,
+                     unsigned int id,
+                     const std::string &description,
+                     const TimeStamp *wake_time,
+                     unsigned int max_burst,
+                     unsigned int rate)
+      : BasicOutputPort(parent, id),
+        m_description(description),
+        m_widget(widget),
+        m_bucket(max_burst, rate, max_burst, *wake_time),
+        m_wake_time(wake_time)
+        {}
+
+  bool WriteDMX(const DmxBuffer &buffer, OLA_UNUSED uint8_t priority) {
+    if (m_bucket.GetToken(*m_wake_time)) {
+      return m_widget->SendDMX(buffer);
+    } else {
+      OLA_INFO << "Port rated limited, dropping frame";
+    }
+    return true;
```

Comment:
While it will make it a little more complicated, you probably want to move the rate limiting check to within OpenDeckWidget.cpp, otherwise you risk wasting your tokens on unchanged packets which don't actually get sent, rather than saving them for the first diff or full packet that arrives!

Depending on your processing times, you might also choose to take another token (I think you can) if processing one (e.g. unpacking the diff) takes longer than the other one on your device.

Reply:
Good point. I've added initial, maybe naive approach. Let me know what you think.

Reply:
Any particular reason you haven't just moved all the TokenBucket code to OpenDeckWidget, rather than keeping it outside and passing it in?

Reply:
I wanted to do that, but to create TokenBucket, I need some parameters first, namely:

    max_burst, rate, max_burst, *wake_time
These aren't accessible in OpenDeckWidget. I might be missing something here.

Reply:
This seems to be a bit beyond my understanding of OLA... Don't know what to do here.

Reply:
Added simpler, alternative approach.